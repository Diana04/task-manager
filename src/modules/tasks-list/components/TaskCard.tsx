import { memo } from 'react';
import { Card, Flex, Tag, Typography } from 'antd';

import { filterTaskTagsByIds, getIsTaskUrgent } from '../helpers';
import { useTasksStore } from '../store';
import type { TagId } from '../types';
import styled from 'styled-components';

const StyledCard = styled(Card)`
    border: none;
    box-shadow: 0 1px 2px 0 rgba(0, 0, 0, 0.03),
        0 1px 6px -1px rgba(0, 0, 0, 0.02), 0 2px 4px 0 rgba(0, 0, 0, 0.02);
    border-left: ${props =>
        props.variant === 'outlined' ? '4px solid #ff7875' : 'none'};
`;

type Props = {
    title: string;
    tagIds: TagId[];
};

export const TaskCard = memo(({ title, tagIds }: Props) => {
    const tags = useTasksStore(state => state.tags);
    const cardTags = filterTaskTagsByIds(tagIds, tags);
    const isUrgent = getIsTaskUrgent(tagIds);

    return (
        <StyledCard size="small" variant={isUrgent ? 'outlined' : 'borderless'}>
            <Flex vertical gap="small">
                <Typography.Title level={5} style={{ margin: 0 }}>
                    {title}
                </Typography.Title>

                {cardTags?.length > 0 && (
                    <Flex gap="4px 0" wrap>
                        {cardTags.map(tag => (
                            <Tag
                                color={tag.color}
                                key={tag.id}
                                bordered={false}
                            >
                                {tag.title}
                            </Tag>
                        ))}
                    </Flex>
                )}
            </Flex>
        </StyledCard>
    );
});
