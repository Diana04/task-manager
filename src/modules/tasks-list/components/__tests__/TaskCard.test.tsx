import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import { TaskCard } from '../TaskCard';

jest.mock('../../helpers', () => ({
    filterTaskTagsByIds: () => [
        {
            id: 'tag-1',
            title: 'tag-1',
            color: 'color-1',
        },
    ],
    getIsTaskUrgent: () => true,
}));

jest.mock('../../store', () => ({
    useTasksStore: (selector: any) =>
        selector({
            tags: [
                {
                    id: 'tag-1',
                    title: 'tag-1',
                    color: 'color-1',
                },
                {
                    id: 'tag-2',
                    title: 'tag-2',
                    color: 'color-2',
                },
            ],
        }),
}));

const setupComponent = (render: Function, props: any = {}) => {
    const fullProps = {
        title: 'title',
        tagIds: ['tag-1'],
        ...props,
    };

    const instance = render(<TaskCard {...fullProps} />);

    return {
        instance,
        props: fullProps,
    };
};

describe('TaskCard', () => {
    describe('snapshot check', () => {
        it('should render correctly according to snapshot', () => {
            const { instance } = setupComponent(render, {});
            expect(instance.container).toMatchSnapshot();
        });
    });

    describe('props check', () => {
        it('should render title', () => {
            const title = 'task-card-title';
            setupComponent(render, { title });
            const titleTextElement = screen.getByText(title);

            expect(titleTextElement).toBeInTheDocument();
        });

        it('should render tags', () => {
            setupComponent(render, {});
            const tagElement = screen.getByText('tag-1');

            expect(tagElement).toBeInTheDocument();
        });
    });
});
