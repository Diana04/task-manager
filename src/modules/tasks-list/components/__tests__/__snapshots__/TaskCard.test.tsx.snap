// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`TaskCard snapshot check should render correctly according to snapshot 1`] = `
<div>
  <div
    class="ant-card ant-card-bordered ant-card-small sc-bRKDuR tLpHK css-dev-only-do-not-override-l9pxc0"
  >
    <div
      class="ant-card-body"
    >
      <div
        class="ant-flex css-dev-only-do-not-override-l9pxc0 ant-flex-align-stretch ant-flex-gap-small ant-flex-vertical"
      >
        <h5
          class="ant-typography css-dev-only-do-not-override-l9pxc0"
          style="margin: 0px;"
        >
          title
        </h5>
        <div
          class="ant-flex css-dev-only-do-not-override-l9pxc0 ant-flex-wrap-wrap"
          style="gap: 4px 0;"
        >
          <span
            class="ant-tag ant-tag-has-color ant-tag-borderless css-dev-only-do-not-override-l9pxc0"
          >
            tag-1
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
